source/darwin/dispatch_queue_event_loop.c

```diff
@@ -0,0 +1,698 @@
+
+    s_lock_cross_thread_data(dispatch_loop);
+    if (!dispatch_loop->synced_data.suspended) {
+        dispatch_loop->synced_data.suspended = true;
+        AWS_LOGF_INFO(AWS_LS_IO_EVENT_LOOP, "id=%p: Stopping event-loop thread.", (void *)event_loop);
+        /* Suspend will increase the dispatch reference count. It is required to call resume before
+         * releasing the dispatch queue. */
+        dispatch_suspend(dispatch_loop->dispatch_queue);
+    }
+    s_unlock_cross_thread_data(dispatch_loop);
+
+    return AWS_OP_SUCCESS;
+}
+
+/**
+ * The function decides if we should run this iteration.
+ * Returns true if we should execute an iteration, false otherwise
+ *
+ * The function should be wrapped with dispatch_loop->context.lock to retain the dispatch loop while running.
+ */
+static bool begin_iteration(struct scheduled_service_entry *entry) {
+    struct dispatch_loop *dispatch_loop = entry->dispatch_queue_context->io_dispatch_loop;
+
+    if (!dispatch_loop) {
+        return false;
+    }
+    return true;
+}
+
+/**
+ * Clean up the related resource and determine if we should schedule next iteration.
+ * The function should be wrapped with dispatch_loop->context.lock to retain the dispatch loop while running.
+ * */
+static void end_iteration(struct scheduled_service_entry *entry) {
+
+    struct dispatch_loop_context *context = entry->dispatch_queue_context;
+    struct dispatch_loop *dispatch_loop = context->io_dispatch_loop;
+
+    s_lock_cross_thread_data(dispatch_loop);
+    dispatch_loop->synced_data.is_executing = false;
+
+    // Remove the node before do scheduling so we didnt consider the entry itself
+    s_lock_service_entries(context);
+    aws_priority_queue_remove(&context->scheduling_state.scheduled_services, entry, &entry->priority_queue_node);
+    s_unlock_service_entries(context);
+
+    bool should_schedule = false;
+    uint64_t should_schedule_at_time = 0;
+    if (!aws_linked_list_empty(&dispatch_loop->synced_data.cross_thread_tasks)) {
+        should_schedule = true;
+    }
+    /* we already know there are tasks to be scheduled, we just want the next run time. */
+    else if (aws_task_scheduler_has_tasks(&dispatch_loop->scheduler, &should_schedule_at_time)) {
+        should_schedule = true;
+    }
+
+    if (should_schedule) {
+        s_lock_service_entries(context);
+        s_try_schedule_new_iteration(context, should_schedule_at_time);
+        s_unlock_service_entries(context);
+    }
+
+    s_unlock_cross_thread_data(dispatch_loop);
+}
+
+// Iteration function that scheduled and executed by the Dispatch Queue API
+static void s_run_iteration(void *context) {
+    struct scheduled_service_entry *entry = context;
+    struct dispatch_loop_context *dispatch_queue_context = entry->dispatch_queue_context;
+    s_acquire_dispatch_loop_context(dispatch_queue_context);
+    s_rlock_dispatch_loop_context(dispatch_queue_context);
+
+    if (!begin_iteration(entry)) {
```

Comment:
The `begin_iteration()` function is only used in this one spot. It might be cleaner and easier to follow if we move the internals of that check (it looks to be checking whether the entry->dispatch_queue_context->io_dispatch_loop is NULL).

Also, instead of `goto iteration_done`, since there's only one spot where we use `iteration_done` we can execute the iteration_done code block here directly and return.